pytorch_lightning/trainer/data_loading.py

```diff
@@ -117,44 +117,69 @@ def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
+
+        - Injecting a `DistributedDataSampler` into the `DataLoader` if on a distributed environment
+        - Wrapping the datasets and samplers into fault-tolerant components
+        """
         if isinstance(dataloader, CombinedLoader):
-            # apply `auto_add_sampler` on all the collection of loaders
+            # apply `prepare_dataloader` on all the collection of loaders
             dataloader.loaders = apply_to_collection(
-                dataloader.loaders, DataLoader, self.auto_add_sampler, shuffle, mode=mode
+                dataloader.loaders, DataLoader, self.prepare_dataloader, shuffle, mode=mode
             )
             return dataloader
 
         # don't do anything if it's not a dataloader
         if not isinstance(dataloader, DataLoader):
             return dataloader
 
+        # the DataLoader should be re-created only if we need to inject
+        # the fault tolerant components or the distributed sampler or we need to track indices for predictions.
         if (
-            self.accelerator_connector.replace_sampler_ddp
-            and self.accelerator_connector.is_distributed
-            and not isinstance(dataloader.sampler, DistributedSampler)
-            and not has_iterable_dataset(dataloader)
+            _fault_tolerant_training()
+            or self._requires_distributed_sampler(dataloader)
+            or mode == RunningStage.PREDICTING
         ):
+            sampler = self._determine_sampler(dataloader, shuffle=shuffle, mode=mode)
+            dataloader = self._prepare_dataloader(dataloader, sampler, mode=mode)
+
+        return dataloader
+
+    def _determine_sampler(self, dataloader: DataLoader, shuffle: bool, mode: Optional[RunningStage] = None) -> Sampler:
+        if self._requires_distributed_sampler(dataloader):
             if not isinstance(dataloader.sampler, (SequentialSampler, RandomSampler)):
                 raise MisconfigurationException(
                     "You seem to have configured a sampler in your DataLoader. This will be replaced "
                     " by `DistributedSampler` since `replace_sampler_ddp` is True and you are using"
                     " distributed training. Either remove the sampler from your DataLoader or set"
-                    " `replace_sampler_ddp`=False if you want to use your custom sampler."
+                    " `replace_sampler_ddp=False` if you want to use your custom sampler."
                 )
-            sampler = self._get_distributed_sampler(dataloader, shuffle, mode=mode)
-        else:
-            # use current sampler
-            sampler = dataloader.sampler
+            return self._get_distributed_sampler(
+                dataloader, shuffle, mode=mode, overfit_batches=self.overfit_batches, **self.distributed_sampler_kwargs
+            )
 
-        dataloader = self.replace_sampler(dataloader, sampler, mode=mode)
-
-        return dataloader
+        return dataloader.sampler
```

Comment:
`_determine_sampler` basically helps to generate distributed_sampler, so there's no need to return `dataloader.sampler`. If the dataloader does not require distributed_sampler, we should not update dataloader again.

Reply:
`_determine_sampler` doesn't generate the distributed sampler. It resolve the right sampler. Changed it to `_resolve_sampler`.